quinn-proto/src/frame.rs

```diff
@@ -922,8 +933,86 @@ impl AckFrequency {
     }
 }
 
+/* Address Discovery https://datatracker.ietf.org/doc/draft-seemann-quic-address-discovery/ */
+
+/// Conjuction of the information contained in the address discovery frames
+/// ([`FrameType::OBSERVED_IPV4_ADDR`], [`FrameType::OBSERVED_IPV6_ADDR`]).
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub(crate) struct ObservedAddr {
+    /// Monotonically increasing integer within the same connection.
+    pub(crate) seq_no: VarInt,
+    /// Reported observed address.
+    pub(crate) ip: IpAddr,
+    /// Reported observed port.
+    pub(crate) port: u16,
+}
+
+impl ObservedAddr {
+    pub(crate) fn new(remote: SocketAddr, seq_no: VarInt) -> Self {
+        Self {
+            ip: remote.ip(),
+            port: remote.port(),
+            seq_no,
+        }
+    }
+
+    /// Get the [`FrameType`] for this frame.
+    pub(crate) fn get_type(&self) -> FrameType {
+        if self.ip.is_ipv6() {
+            FrameType::OBSERVED_IPV6_ADDR
+        } else {
+            FrameType::OBSERVED_IPV4_ADDR
+        }
+    }
+
+    /// Compute the number of bytes needed to encode the frame.
+    pub(crate) fn size(&self) -> usize {
+        let type_size = VarInt(self.get_type().0).size();
+        let req_id_bytes = self.seq_no.size();
+        let ip_bytes = if self.ip.is_ipv6() { 16 } else { 4 };
+        let port_bytes = 2;
+        type_size + req_id_bytes + ip_bytes + port_bytes
+    }
+
+    /// Unconditionally write this frame to `buf`.
+    pub(crate) fn write<W: BufMut>(&self, buf: &mut W) {
+        buf.write(self.get_type());
+        buf.write(self.seq_no);
+        match self.ip {
+            IpAddr::V4(ipv4_addr) => {
+                buf.write(ipv4_addr);
+            }
+            IpAddr::V6(ipv6_addr) => {
+                buf.write(ipv6_addr);
+            }
+        }
+        buf.write::<u16>(self.port);
+    }
+
+    /// Reads the frame contents from the buffer.
+    ///
+    /// Should only be called when the fram type has been identified as
+    /// [`FrameType::OBSERVED_IPV4_ADDR`] or [`FrameType::OBSERVED_IPV6_ADDR`].
+    pub(crate) fn read<R: Buf>(bytes: &mut R, is_ipv6: bool) -> coding::Result<Self> {
+        let seq_no = bytes.get()?;
+        let ip = if is_ipv6 {
+            IpAddr::V6(bytes.get()?)
+        } else {
+            IpAddr::V4(bytes.get()?)
+        };
+        let port = bytes.get()?;
+        Ok(Self { seq_no, ip, port })
+    }
+
+    /// Gives the [`SocketAddr`] reported in the frame.
```

Comment:
This project's convention is to have the first line of a doc comment not end with a period